src/hotspot/share/gc/shared/c2/barrierSetC2.cpp

```diff
@@ -108,6 +109,33 @@ Label* BarrierStubC2::continuation() {
   return &_continuation;
 }
 
+void BarrierStubC2::preserve(Register r) {
+  const VMReg vm_reg = r->as_VMReg();
+  assert(vm_reg->is_Register(), "r must be a general-purpose register");
+  _preserve.Insert(OptoReg::as_OptoReg(vm_reg));
+}
+
+void BarrierStubC2::dont_preserve(Register r) {
+  const VMReg vm_reg = r->as_VMReg();
+  assert(vm_reg->is_Register(), "r must be a general-purpose register");
+  OptoReg::Name reg = OptoReg::as_OptoReg(vm_reg);
+  // Subtract not only reg, but also all related OptoRegs that are sub-registers
+  // of the same general-purpose, processor register (e.g. {R11, R11_H} for r11
+  // in aarch64). We assume that all related OptoRegs have contiguous indices.
+  while (OptoReg::is_reg(reg)) {
+    const VMReg vm_reg = OptoReg::as_VMReg(reg);
+    if (!(vm_reg->is_Register()) || vm_reg->as_Register() != r) {
```

Comment:
This doesn't work on PPC64: We run into "assert(is_Register() && is_even(value())) failed: even-aligned GPR name" (vmreg_ppc.hpp:54). Calling `as_Register()` is only supported for the even ones.

Reply:
Thank you Martin for trying out the patch and for the suggestion, will test a solution based on `is_concrete()` and push the changes later, if it works.

Reply:
Done (commit https://github.com/openjdk/jdk/pull/19026/commits/c0fc66deb654a9b930a7b7cf1a7e7fa093739027). @TheRealMDoerr please let me know if this works on PPC64.

Reply:
Yes, this works on PPC64. Thanks!